packages/loopring_v3/circuit/Gadgets/StorageGadgets.h

```diff
@@ -199,47 +215,49 @@ class StorageReaderGadget : public GadgetT {
   }
 };
 
-class NonceGadget : public GadgetT {
+// QUESTION(daniel): it seems this gadget is the same as ReadStorageGadget
+// except that it requires data to be 0.
```

Comment:
Yes, that's how the nonce system on top of the storage works. The storage data needs to be 0 (or the storage is overwritten). Afterwards the the slot will contain 1, which marks the transfer as consumed so it cannot be reused again.

Reply:
OK, I'll add some more information as comments for the auditors.